Add support for x86_32, arm, and arm64 architectures

tyhicks opened the pull request:
(This is a continuation of #9)

Some hardening recommendations are dependent on the processor architecture. For example, the KSPP recommendations differ between [x86_32](https://kernsec.org/wiki/index.php/Kernel_Self_Protection_Project/Recommended_Settings#x86_32) and [x86_64](https://kernsec.org/wiki/index.php/Kernel_Self_Protection_Project/Recommended_Settings#x86_64).

This pull request adds the ability to reason about the architecture when constructing the checklist. It also teaches the script about `x86_32`, `arm`, and `arm64` specific config recommendations.

I verified that all the example configs in `config_files/` show the same number of config check failures before and after these changes are applied. Of course, the ordering of the options are changed since the ordering used to construct the checklist has been changed.

Some changes since #9 include:
- Drop kernel version detection from the pull request
- Rename `detect_arch_and_version()` to `detect_arch_from_config()`
- Look for `CONFIG_X86_32` and `CONFIG_X86_64` when detecting `x86` sub architecture
- Restrict the accepted `-a <ARCHITECTURE>` values to those found in `SUPPORTED_ARCHS`

a13xp0p0v commented:
Hello @tyhicks , thanks a lot for the follow-up! Let me propose some improvements.

a13xp0p0v commented:
@tyhicks , thanks for your work again!
Let me propose one more idea. What do you think about splitting [KSPP recommended settings](http://kernsec.org/wiki/index.php/Kernel_Self_Protection_Project/Recommended_Settings) onto 4 arch-specific configs in `./configs/`?

tyhicks commented:
Yes, I can add 4 arch-specific configs in `./config_files/`.

tyhicks commented:
I've rebased on top of your current tree, fixed up a few things, added what I think you were asking for in the arch-specific KSPP files, and force pushed to this branch.

a13xp0p0v commented:
Ouch. 
@tyhicks , excuse me please!
I've made a code review 3 days ago, but didn't hit "submit" button, so it is "pending" :(

a13xp0p0v commented on kconfig-hardened-check.py:
```diff
@@ -310,9 +323,21 @@ def check_config_file(fname):
     parser.add_argument('--debug', action='store_true', help='enable internal debug mode')
     args = parser.parse_args()
 
+    if not args.arch:
```
If you don't have time/desire, I can pick up your branch and polish it myself.
Thank you again!

tyhicks replied:
With these changes, I no longer see a need for the `-a` flag except when also specifying the `-p` flag with no config file (`-c`).

Also, I don't feel like printing config recommendations for all architecture is useful when `-a` is not specified.

a13xp0p0v commented:
If you don't have time/desire, I can pick up your branch and polish it myself.
Thank you again!

tyhicks commented on kconfig-hardened-check.py:
```diff
@@ -34,8 +34,9 @@
+X86_32 = "x86_32"
 X86_64 = "x86_64"
-SUPPORTED_ARCHS = [ X86_64 ]
+SUPPORTED_ARCHS = [ X86_32, X86_64 ]
```
I capitalized it since it is a "constant". That's what PEP 8 [suggests](https://www.python.org/dev/peps/pep-0008/#constants).

Renaming it to match the other globals also makes sense, though.

tyhicks commented on kconfig-hardened-check.py:
```diff
@@ -111,31 +112,42 @@ def check(self):
+                arch = match.group('arch')
+                break
+
+        if arch == 'x86':
```
> I think we can avoid parsing the comment "# Linux/(?P\w+) \S+ Kernel Configuration".

Yeah, that makes sense.

> So we even can avoid things like `X86_32 = "x86_32"` above and just store strings in supported_archs.

I don't follow you here. We still need some way to convey which architecture we're dealing with when, for instance, constructing the checklist. We could hardcode the architecture strings there but I don't see what that gains from having variables that we can refer to.

tyhicks commented:
> If you don't have time/desire, I can pick up your branch and polish it myself.

I won't mind if you do the polishing yourself.

> Thank you again!

No problem. Thanks for all the review comments.

a13xp0p0v commented:
Hello @tyhicks ,

I've finished with arch support based on your work.
Do you like it?
Do you have any comments or requests?
Thanks!

tyhicks commented:
Thanks for finishing out the work. It looks very good to me. I'll make use of the changes over the next week or so and submit new pull requests if I spot anything wrong/missing. Thanks again!
